jupyterlab_bigquery/jupyterlab_bigquery/pagedAPI_handler/pagedAPI_handler.py

```diff
@@ -0,0 +1,123 @@
+from enum import Enum
+from notebook.base.handlers import APIHandler, app_log
+from logging import INFO, WARN
+import json
+from threading import Lock
+from collections import deque, defaultdict
+from abc import ABC, abstractmethod
+
+START_STATE = 'start'
+CONTINUE_STATE = 'continue'
+CANCEL_STATE = 'cancel'
+
+class PagedAPIHandler(APIHandler, ABC):
+  '''
+    Enables easy to use paged/batched API requests between front end and backend.
+    User needs to implement query and cancel to provide a generator and the ability
+    to cancel a job (if necessary).  Exceptions can be safely raised from these method,
+    they will will be sent to front end.
+  '''
+
+  # job_id: deque[(generator, job_obj)], remove after done
+  generator_pool = defaultdict(lambda: None)
+  generator_lock = Lock()
+
+  def post(self, *args, **kwargs):
+    '''
+      Entry point for API. Handles start, continue, and cancel.
+    '''
+    post_body = self.get_json_body()
+
+    intention = post_body['intention']
+    load = post_body['load']
+
+    if intention == START_STATE:
+        self._onStart(load)
+    elif intention == CONTINUE_STATE:
+        self._onContinue(load)
+    elif intention == CANCEL_STATE:
+        self._onCancel(load)
+
+  def _onStart(self, load):
+      request_body = load['requestBody']
+      page_size = load['pageSize']
+      query_generator = self.query(request_body, page_size)
+
+      id = None
+      error = None
+      try:
+        job, id = next(query_generator)
+      except Exception as err:
+        error = str(err)
+        app_log.log(WARN, 'Failed started query: %s', error)
+      
+      if error is None:
+        with PagedAPIHandler.generator_lock:
+          PagedAPIHandler.generator_pool[id] = query_generator, job
+          app_log.log(INFO, 'Successfully started query %s', id)
+
+      self.finish({'id': json.dumps(id), 'error': json.dumps(error)})
+
+  def _onContinue(self, load):
+      id = load['id']
+      query_generator = None
+
+      with self.generator_lock:
+          query_generator, _ = PagedAPIHandler.generator_pool[id]
+
+      finish = False
+      load = []
+      error = None
+      if query_generator is None:
+          finish = True
+      else:
+          try:
+            load = next(query_generator)
+          except StopIteration:
+            finish = True
+            with PagedAPIHandler.generator_lock:
+                del PagedAPIHandler.generator_pool[id]
+            app_log.log(INFO, 'Successfully finished query %s', id)
```

Comment:
We might should think about clearing out generators after a timeout period, but I don't it is necessary to start.